### PR TITLE
Upgrade mapbox-gl to 0.43.0

### LINKED ISSUE #493: Error removing layer when using mapbox-gl 0.43.0
Mapbox-gl v0.43.0 introduces a breaking change that throws an error if a source is removed while still in use (e.g. remove a source prior to removing layers that use the source, https://github.com/mapbox/mapbox-gl-js/pull/5562 ). This was discovered in #491.

Example error:
```
mapbox-gl.js:546 Error: Source "example_id" cannot be removed while layer "example_id_marker" is using it.
    at t.removeSource (mapbox-gl.js:406)
    at e.removeSource (mapbox-gl.js:520)
    at Source.componentWillUnmount (source.js:71)
    at callComponentWillUnmountWithTimerInDev (react-dom.development.js:11117)
    at HTMLUnknownElement.callCallback (react-dom.development.js:1299)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:1338)
    at invokeGuardedCallback (react-dom.development.js:1195)
    at safelyCallComponentWillUnmount (react-dom.development.js:11125)
    at commitUnmount (react-dom.development.js:11415)
    at unmountHostComponents (react-dom.development.js:11356)
```

### LINKED ISSUE #499: Example fails to build due to typescript errors
Example does not build after commit d9359a15ee6ebe6afec759bcc6e0edae54289fdc . Various TS2403 and TS2605 errors reported.

[typescript.txt](https://github.com/alex3165/react-mapbox-gl/files/1649254/typescript.txt)

### PR DESCRIPTION
**Pull request by jthetzel**
1) Upgrade mapbox-gl to 0.43.0
2) Remove layers prior to removing source fixes #493 
3) Upgrade react, react-dom in example fixes #499 
4) Added toggleLayer demo to demonstrate #493 is now fixed

### CODE REVIEW

**Comment by jthetzel** [comment]
There still might be an error when a source receives a new tile source ( https://github.com/alex3165/react-mapbox-gl/blob/d9359a15ee6ebe6afec759bcc6e0edae54289fdc/src/source.ts#L110 ). Related issues: 
https://github.com/mapbox/mapbox-gl-js/issues/5879
https://github.com/mapbox/mapbox-gl-js/issues/2633

**Comment by alex3165** [comment]
Hey @jthetzel Thank you for the PR it is very appreciated, would you mind to just add a `toggle` button for the layer to the switch style  and maybe rename the name of the example to `toggle-layer-style`

**Comment by Wykks** [comment]
I'm not sure If removing layers when removing the source is a good solution. This "magic" could lead to weird situations without errors in console.
But, that's better than nothing :man_shrugging:

File: src/geojson-layer.ts [hunk 1]
```diff
@@ -205,6 +205,14 @@ export default class GeoJSONLayer extends React.Component<Props> {
     const { map } = this.context;
 
     if (map.getSource(this.id)) {
+      const { layers } = map.getStyle()
+
+      if (layers) {
+        layers
+          .filter(layer => layer.source === this.id)
+          .map(layer => map.removeLayer(layer.id))
```
Comment (alex3165) [inline src/geojson-layer.ts]: Can you use `forEach` instead of `map` please?

File: src/source.ts [hunk 2]
```diff
@@ -89,6 +89,14 @@ export default class Source extends React.Component<Props> {
     map.off('styledata', this.onStyleDataChange);
 
     if (map.getSource(this.id)) {
+      const { layers } = map.getStyle()
+
+      if (layers) {
+        layers
+          .filter(layer => layer.source === this.id)
+          .map(layer => map.removeLayer(layer.id))
```
Comment (alex3165) [inline src/source.ts]: Same if you could use `forEach` that would be great

**Comment by alex3165** [comment]
@Wykks I agree, we should have this specified on the documentation for `Source` and `GeoJSONLayer` components, something like:
When unmounting sources, react-mapbox-gl will first remove the related layers therefore this can lead to inconsistent state with your `Layer` component state being still mounted but with the actual layer not being displayed on the map.

**Comment by alex3165** [comment]
@jthetzel If you could update the documentation with the above description and combine your example with the `switch-style` example then I will make sure to get this merged asap. Thanks

**Comment by jthetzel** [comment]
I agree with @Wykks that the layer removal is sub-optimal. However, I don't yet see an elegant way to translate your ngZone buffer to react-mapbox-gl without a significant architectural change.

**Comment by jthetzel** [comment]
I should have time to get this done tonight (NST time).

**Comment by alex3165** [comment]
Closing this PR in favor of https://github.com/alex3165/react-mapbox-gl/pull/507 which include those changes + some improvements

**Pull request closed without merging**